build.rs

```diff
@@ -1,10 +1,116 @@
-use std::env;
+use regex::Regex;
+use std::fmt::{Display, Formatter};
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::{env, fmt};
+
+#[derive(Clone, Debug)]
+pub struct Target {
+    pub architecture: String,
+    pub vendor: String,
+    pub system: Option<String>,
+    pub abi: Option<String>,
+}
+
+impl Target {
+    pub fn as_strs(&self) -> (&str, &str, Option<&str>, Option<&str>) {
+        (
+            self.architecture.as_str(),
+            self.vendor.as_str(),
+            self.system.as_deref(),
+            self.abi.as_deref(),
+        )
+    }
+}
+
+impl Display for Target {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "{}-{}", &self.architecture, &self.vendor)?;
+
+        if let Some(ref system) = self.system {
+            write!(f, "-{}", system)
+        } else {
+            Ok(())
+        }?;
+
+        if let Some(ref abi) = self.abi {
+            write!(f, "-{}", abi)
+        } else {
+            Ok(())
+        }
+    }
+}
+
+pub fn ndk() -> String {
+    env::var("ANDROID_NDK").expect("ANDROID_NDK variable not set")
+}
+
+pub fn target_arch(arch: &str) -> &str {
+    match arch {
+        "armv7" => "arm",
+        "aarch64" => "arm64",
+        "i686" => "x86",
+        arch => arch,
+    }
+}
+
+fn host_tag() -> String {
+    // Because this is part of build.rs, the target_os is actually the host system
+    if cfg!(target_os = "windows") {
+        "windows-x86_64"
+    } else if cfg!(target_os = "linux") {
+        "linux-x86_64"
+    } else if cfg!(target_os = "macos") {
+        "darwin-x86_64"
+    } else {
+        panic!("host os is not supported")
```

Comment:
I don't know much about android so it's hard for me to know if this is right or wrong, but I do know that sometimes I am frustrated by proactive checks like this, when I'm on a system that's close enough to one of these where things will still work, but don't specifically because of the check. that's only something you can really know though, and maybe you just wouldn't want to support those cases anyway.

Reply:
Honestly not much can be done here as [they](https://developer.android.com/ndk/guides/other_build_systems#overview) only support macOS, Linux and Windows